mm/2s2h/ActorExtension/ActorExtension.cpp

```diff
@@ -0,0 +1,106 @@
+
+#include "ActorExtension.h"
+#include "2s2h/ShipInit.hpp"
+#include "2s2h/GameInteractor/GameInteractor.h"
+
+ActorExtensionId sNextActorExtensionId = 0;
+
+struct ActorExtensionKeyHash {
+    std::size_t operator()(const std::pair<Actor*, ActorExtensionId>& key) const {
+        return std::hash<Actor*>{}(key.first) ^ (std::hash<ActorExtensionId>{}(key.second) << 1);
+    }
+};
+
+std::unordered_map<ActorExtensionId, size_t> sGlobalActorExtensionSizes;
+std::unordered_map<s16, std::unordered_map<ActorExtensionId, size_t>> sActorExtensionSizes;
+std::unordered_map<std::pair<Actor*, ActorExtensionId>, void*, ActorExtensionKeyHash> sActorExtensionData;
+
+void* ActorExtension_Get(Actor* actor, ActorExtensionId id) {
+    if (actor == NULL) {
+        return NULL;
+    }
+
+    auto it = sActorExtensionData.find(std::make_pair(actor, id));
+    if (it == sActorExtensionData.end()) {
+        return NULL;
+    }
+
+    return it->second;
+}
+
+ActorExtensionId ActorExtension_CreateForId(s16 actorId, size_t size) {
+    ActorExtensionId id = sNextActorExtensionId++;
+    sActorExtensionSizes[actorId][id] = size;
+    return id;
+}
+
+ActorExtensionId ActorExtension_CreateForAll(size_t size) {
+    ActorExtensionId id = sNextActorExtensionId++;
+    sGlobalActorExtensionSizes[id] = size;
+    return id;
+}
+
+void ActorExtension_Alloc(Actor* actor, s16 actorId) {
+    if (actor == NULL) {
+        return;
+    }
+
+    for (auto& [id, size] : sGlobalActorExtensionSizes) {
+        void* data = malloc(size);
+        memset(data, 0, size);
+        sActorExtensionData[std::make_pair(actor, id)] = data;
+    }
+
+    for (auto& [id, size] : sActorExtensionSizes[actorId]) {
+        void* data = malloc(size);
+        memset(data, 0, size);
+        sActorExtensionData[std::make_pair(actor, id)] = data;
+    }
+}
+
+void ActorExtension_Free(Actor* actor) {
+    if (actor == NULL) {
+        return;
+    }
+
+    for (auto& [id, size] : sGlobalActorExtensionSizes) {
+        auto it = sActorExtensionData.find(std::make_pair(actor, id));
+        if (it != sActorExtensionData.end()) {
+            free(it->second);
+            sActorExtensionData.erase(it);
+        }
+    }
+
+    for (auto& [id, size] : sActorExtensionSizes[actor->id]) {
+        auto it = sActorExtensionData.find(std::make_pair(actor, id));
+        if (it != sActorExtensionData.end()) {
+            free(it->second);
+            sActorExtensionData.erase(it);
+        }
+    }
+}
+
+// Example usage, the storing of the actorListIndex happens in Actor_SpawnAsChildAndCutscene:
+
+ActorExtensionId actorListIndexActorExt;
+s16 currentActorListIndex = 0;
+
+static RegisterShipInitFunc initFunc(
+    []() {
+        HOOK(OnPortInit, []() { // Only ever called once
+            actorListIndexActorExt = ActorExtension_CreateForAll(sizeof(s16));
+        });
+    },
+    {});
```

Comment:
This feels redundant. Why create a `OnPortInit` when `ShipInit::InitAll` is called just before this anyways? It leaves two long living callbacks registered, instead of one.

Also the `HOOK()` does not unregister, so if `ShipInit::InitAll` is executed again, that causes there to be multiple+ hooks registered (even if nothing will ever trigger them).

Regardless, this should have a static var involved in some fashion to prevent multiple registers, or even just reuse the `currentActorListIndex` id itself.

My suggestion would be to drop `OnPortInit` entirely and just write the ShipInit like this:

```suggestion
ActorExtensionId actorListIndexActorExt = 0;
s16 currentActorListIndex = 0;

static RegisterShipInitFunc initFunc(
    []() {
        if (actorListIndexActorExt == 0) {
            actorListIndexActorExt = ActorExtension_CreateForAll(sizeof(s16));
        }
    },
    {});
```